### ISSUE TITLE
`convert` fails to find valid conversion

### ISSUE THREAD

**Issue by gwhitney** [josdejong/typed-function#128]
Consider the type system
```
typed.types = [
  { name: 'number',    test: x => typeof x === 'number' },
  {
    name: 'identifier',
    test: x => (typeof x === 'string' &&
                /^\p{Alphabetic}[\d\p{Alphabetic}]*$/u.test(x))
  },
  { name: 'string',    test: x => typeof x === 'string' }
]

typed.addConversion({ from: 'string', to: 'number', convert: x => parseFloat(x) })

const check = typed('check', {
  identifier: i => 'found an identifier: ' + i,
  string: s => s + ' is just a string'
})

console.log(check('xy33')) // returns 'found an identifier xy33'
console.log(check('Wow!')) // returns 'Wow! is just a string'

typed.convert('123.5', 'number')
typed.convert('Infinity', 'number')
```

All goes fine until the last line when typed-function throws `Error: Cannot convert from identifier to number` even though of course the conversion from 'Infinity' to a floating point number is perfectly valid. I would consider this a bug (since otherwise overlapping types works fine in other typed-function facilities).

**Comment by josdejong** [comment]
Thanks, that is indeed a bug. 

I suppose this issue is related to `identifier` _also_ being a `string`.

**Comment by gwhitney** [comment]
> I suppose this issue is related to `identifier` _also_ being a `string`.

Indeed. In fact, one approach to fixing it would be allowing a type to declare it `isa` some other type, and fully handling subtype/supertype hierarchies in typed-function.js. That change of course eliminates this bug, and removes redundancy in type checks (If `identifier` includes `isa: 'string'` then its test function does not have to test its argument is a string, that will already
have been done by the time its test is executed). It also allows rudimentary intersection types as well, if you allow the value of `isa` to be a list of types rather than only a single type. And in fact, I have a PR close to done for such a change. But before I take the time to finish up that PR -- is that too significant a change for typed-function for you to contemplate merging at this juncture? It does affect most of the methods in typed-function.js. Thanks for letting me know.

**Comment by gwhitney** [comment]
Oh and note adding subtypes/supertypes also means that the type identification aspect mentioned in #127 becomes "complete" as well: the type identifier of a value becomes all of the minimal types it inhabits.

**Comment by josdejong** [comment]
Yeah. The issue boils down to the possibility that one value can match multiple types, the library doesn't reckon with that. I think to fix this bug, we can change the `convert` function from first determining a single type for the value at hand into either creating a util function that returns all matching types instead of one, or simply looping over all conversions, an for each conversion check if the actual value matches the `from` type of the conversion. Which approach performs best depends on the number of conversions vs the number of types I guess. We sort of already have a list with `isa` functions, the `types`. For efficiency we may want to have them in a map/object to be able to look them up quickly by type name (but let's be careful with assumptions in this regard).

**Comment by josdejong** [comment]
I think in general we could improve the performance of `typed-function` by letting `findTypeByName` use a lookup table instead of iterating over an array each and every time... that would be interesting to try out!

**Comment by gwhitney** [comment]
> we can change the `convert` function from first determining a single type for the value at hand into either creating a util function that returns all matching types instead of one, or simply looping over all conversions, an for each conversion check if the actual value matches the `from` type of the conversion. Which approach performs best depends on the number of conversions vs the number of types I guess. We sort of already have a list with `isa` functions, the `types`.

Right, that's certainly how I started, by looping through all the types. Then I noticed that in the main use case I have access to, some of the types in the list are in fact subtypes of others. So then the redundant testing when going through a long list of types seemed wasteful, and plus I thought it would make the type system clearer to officially designate that some types in the list are subtypes of others.

So to perhaps make it clearer, what I am proposing would allow something like
```
typed.addType({name: 'string', test: x => typeof x === `string`})
typed.addType({name: 'identifier', isa: 'string', test: x => /^[a-zA-Z_]\w*$/.test(x)})
```
And then if you defined a conversion from string to number, it would automatically apply to the identifier "Infinity", because typed-function would know that anything that is an identifier is also a string. And note that the string test is specifically **not** reiterated in the identifier test; because of the 'isa' property, the infrastructure knows to test for being a string first, and only if that succeeds, try calling the identifier test.

I chose the property name 'isa' partly because it is short and sweet, but that could easily be changed if you preferred something like 'subtypeOf' or even 'extends' (which means the same thing but I always find it semantically jarring in my head that if type A extends typed B that means that type A is a subset of type B, I guess because for me "extends" has the connotation of making bigger and "subset" has the connotation of being smaller).

> I think in general we could improve the performance of typed-function by letting findTypeByName use a lookup table instead of iterating over an array each and every time... that would be interesting to try out!

Yes absolutely that change is in the PR I've been working on. If you think there's a chance you might be interested in this extension of the mechanism to define types in typed-function, let me know and I will be happy to finish off the PR and submit it at least for you to look it over (and of course, no hard feelings if you decide not to go that way when you see it). On the other hand, if you're pretty sure it's just not a direction you'd like typed-function to head right now, then I can just drop the project (again, no hard feelings).

**Comment by josdejong** [comment]
I understand your proposal, which would allow to clearly define relations between types. I hope though that there is no need to define these relations: it can be tricky to keep such a mapping up to date I expect. It's easier not to have to think about it at all.

I worked out one way to solve the issue at hand in #132, this does not require an `isa` mapping, and I think it "just" works, but maybe I'm overlooking something. It solves the issue of picking the first matching type and using that to find a conversion and instead tries to find the first conversion that matches any of the matching types.

**Comment by gwhitney** [comment]
OK, I will hold up on working on the "isa" idea at the moment, unless/until a more compelling use case comes up, and look at and comment on your #132 as soon as I have a chance.

**Comment by gwhitney** [comment]
Resolved in v3 (#147)
